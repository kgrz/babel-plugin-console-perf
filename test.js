const babel = require('babel-core');
const plugin = require('./index');

console.profile = jest.fn();
console.profileEnd = jest.fn();

it('wraps a function in profile block', () => {
	const example = `
	const b = () => {
		// profile
		return 42;
	}

	const a = () => {
		// profile
		return b();
	}
	`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('caters to the case of multiple returns', () => {
	const example = `
	const a = () => {
		// profile
		if (true) {
			return Math.random(12);
		}

		return Math.sqrt(10000000000000000000);
	}`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('does not add profile statement if one is already present', () => {
	const example = `
	const b = () => {
		// profile

		console.profile('some name');
		return 42;
		console.profileEnd();
	}`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('has multiple returns, one of them invalid', () => {
	const example = `
	const b = () => {
		// profile

		return 42;
		return 42;
		return 42;
	}`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});


it('has multiple comments', () => {
	let example = `
	const b = () => {
		// profile
		// profile
		// profile
		// profile
		// profile

		return 42;
	}`;

	let code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();

	example = `
	const b = () => {
		// profile
		//
		// profile
		//
		return 42;
		// profile
		//
		// profile

	}`;

	code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('maintains indentation with nested functions', () => {
	let example = `
	const b = () => {
		const a = () => {
			return Math.sqrt(1000000000000);
		}

		// profile
		return 42;
	}`;

	let code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();

	example = `
	const b = () => {
		const a = () => {
			// profile
			return Math.sqrt(1000000000000);
		}

		// profile
		return a();
	}`;

	code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('maintains indentation when used with iterators', () => {
	const example = `
	const b = () => {
		[1, 3, 5].map(item => {
			return item;
		})

		// profile
		return 42;
	}`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});

it('maintains indentation when for if blocks', () => {
	const example = `
	const b = () => {
		// profile
		
		if (true) {
			return 12;
		}

		switch (true) {
			case 'case1':
				return 'what';
			case 'case 2': {
				return 'where';
			}
			case 'case 3':
				return 'who';
			default:
				return 'default'
		}

		return 42;
	};
	`;

	const code = babel.transform(example, { plugins: [ plugin ] }).code;
	expect(code).toMatchSnapshot();
});


describe('usage inside classes', () => {
	it('works inside function properties', () => {
		const example = `
		class A {
			changeVisualState (id) {
				// profile
				return item => {
					this.setState({
						newState: {}
					});
					callSomeMethod();
				};
			}
		}
		`;

		const code = babel.transform(example, { plugins: [ plugin ] }).code;
		expect(code).toMatchSnapshot();
	});
});

