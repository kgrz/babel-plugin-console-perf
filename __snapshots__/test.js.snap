// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`caters to the case of multiple returns 1`] = `
"
const a = () => {
	console.profile(\\"a\\");

	// profile
	if (true) {
		let _babel_temp_alias = Math.random(12);

		console.profileEnd();

		return _babel_temp_alias;
	}

	_babel_temp_alias = Math.sqrt(10000000000000000000);
	console.profileEnd();
	return _babel_temp_alias;
};"
`;

exports[`does not add profile statement if one is already present 1`] = `
"
const b = () => {
	console.profile('b');

	// profile

	console.profile('some name');
	console.profileEnd();
	return 42;
	console.profileEnd();
};"
`;

exports[`has multiple comments 1`] = `
"
const b = () => {
	console.profile(\\"b\\");
	console.profileEnd();

	// profile
	// profile
	// profile
	// profile
	// profile

	return 42;
};"
`;

exports[`has multiple comments 2`] = `
"
const b = () => {
	console.profile(\\"b\\");
	console.profileEnd();

	// profile
	//
	// profile
	//
	return 42;
	// profile
	//
	// profile
};"
`;

exports[`has multiple returns, one of them invalid 1`] = `
"
const b = () => {
	console.profile(\\"b\\");
	console.profileEnd();

	// profile

	return 42;
	console.profileEnd();
	return 42;
	console.profileEnd();
	return 42;
};"
`;

exports[`maintains indentation when for if blocks 1`] = `
"
const b = () => {
	console.profile('b');

	// profile

	if (true) {
		console.profileEnd();

		return 12;
	}

	switch (true) {
		case 'case1':
			console.profileEnd();

			return 'what';
		case 'case 2':
			{
				console.profileEnd();

				return 'where';
			}
		case 'case 3':
			console.profileEnd();

			return 'who';
		default:
			console.profileEnd();

			return 'default';
	}

	console.profileEnd();
	return 42;
};"
`;

exports[`maintains indentation when used with iterators 1`] = `
"
const b = () => {
	console.profile(\\"b\\");

	[1, 3, 5].map(item => {
		return item;
	});

	// profile
	console.profileEnd();
	return 42;
};"
`;

exports[`maintains indentation with nested functions 1`] = `
"
const b = () => {
	console.profile(\\"b\\");

	const a = () => {
		return Math.sqrt(1000000000000);
	};

	// profile
	console.profileEnd();
	return 42;
};"
`;

exports[`maintains indentation with nested functions 2`] = `
"
const b = () => {
	console.profile(\\"b\\");

	const a = () => {
		console.profile(\\"a\\");

		let _babel_temp_alias = Math.sqrt(1000000000000);

		console.profileEnd();

		// profile
		return _babel_temp_alias;
	};

	// profile

	let _babel_temp_alias = a();

	console.profileEnd();
	return _babel_temp_alias;
};"
`;

exports[`usage inside classes works inside function properties 1`] = `
"
class A {
	changeVisualState(id) {
		console.profile(\\"changeVisualState\\");

		let _babel_temp_alias = item => {
			this.setState({
				newState: {}
			});
			callSomeMethod();
			console.profileEnd();
		};

		console.profileEnd();

		// profile
		return _babel_temp_alias;
	}
}"
`;

exports[`wraps a function in profile block 1`] = `
"
const b = () => {
	console.profile(\\"b\\");
	console.profileEnd();

	// profile
	return 42;
};

const a = () => {
	console.profile(\\"a\\");

	let _babel_temp_alias = b();

	console.profileEnd();

	// profile
	return _babel_temp_alias;
};"
`;
